xgbse/metrics.py

```diff
@@ -65,28 +65,41 @@ def concordance_index(y_true, survival, risk_strategy="mean", which_window=None)
     events, times = convert_y(y_true)
     events = events.astype(bool)
 
-    cind_df = pd.DataFrame({"t": times, "e": events, "r": risks})
+    time_order = times.argsort()
+    all_samples = len(hazards)
+
+    times = times[time_order]
+    hazards = hazards[time_order]
+    events = events[time_order]
 
     count_pairs = 0
     concordant_pairs = 0
     tied_pairs = 0
 
     # running loop for each uncensored sample,
     # as by https://arxiv.org/pdf/1811.11347.pdf
-    for _, row in cind_df.query("e == True").iterrows():
-
-        # getting all censored and uncensored samples
-        # after current row
-        samples_after_i = cind_df.query(f"""{row['t']} < t""")
+    for i in range(all_samples):
+        if events[i]:
+            # getting all censored and uncensored samples
+            # after current row
+            pairs, concord, tied = _concordant_pairs(hazards[i + 1 :], hazards[i])
```

Comment:
For the last loop: 
when we get up to hazards[i+1 :], if we get a last element as a True event, do we still have an element with index i+1? Doesn't it raise an error?

Reply:
It doesn't because it's a slice, it'll just be an empty list

Reply:
That's right! Nice point